### ISSUE TITLE
dependency on implicit type conversion in jelly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10410]
|<img alt="enki's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| enki reported|
|-|-|
|Trac ID|trac#555|
|Type|defect|
|Created|2004-03-18 01:34:02Z|
|Branch|https://github.com/twisted/twisted/tree/implicit-jelly-555|

Attachments:

* [555.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cfa/cfa1150f1787186742a9a884b73a43d8cf219f9b/f7140db28a6d9e0762a196eada0bda32077a0b71.patch) (2180 bytes) - added by aruse on 2011-03-14 22:29:22Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__555 555
type__defect defect
reporter__enki enki
priority__low low
milestone__ 
branch__branches_implicit_jelly_555 branches/implicit-jelly-555
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__easy easy
time__1079573642000000 1079573642000000
changetime__1476058428184453 1476058428184453
version__ 
owner__exarkun exarkun
cc__glyph cc__warner cc__enki
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="enki's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|enki commented|
|-|-|

```
#!html
<pre>
jelly.py:511
    def unjelly(self, obj):
        if type(obj) is not types.ListType:
            return obj

flavors.py:131
    return "remote", jellier.invoker.registerReference(self)

Referenceable.jellyFor returns a tuple (remote, someId)
jelly.unjelly checks if this type(tuple) is types.ListType
unjelly returns the tuple, instead of the RemoteReference that would be correct

somewhere in between the tuple is implicitly converted to a list.

at this point i've come to the conclusion that this implicity dependency on a 
type conversion is the reason for death and destruction of small and furry 
animals and a major contributing factor to global warming.

proposed fix: yuck. the whole implicit dependency on the type is error prone.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Sorry for the long lag.  This, and about 9 billion other issues, are going to be fixed in new pb, so it's 
warner's problem.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
hmm. newpb's serialization uses an ISlicer adapter for each object. Tuples are
(for example) serialized because the TupleSlicer class has been registered as
the ISlicer adapter for types.TupleType, so effectively we have a dictionary
indexed by type(obj).

I can't tell if this is better or worse than the way oldjelly did it.

Is there any chance you can put together a test case for this? Then it would be
a simple matter to determine if newpb has fixed the problem or not.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

What is "implicit type conversion"?  Python doesn't do anything like that.  So Twisted must convert the type.  But then it would have to be explicit?

This bug report makes no sense.  Where is the bug being reported?  Is something broken?  Is there a failing test case?  Is there a failing program?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun|
|-|-|

Okay, here's the failure:

```
>>> from twisted.spread.pb import Broker
>>> from twisted.spread.flavors import Referenceable
>>> from twisted.spread.jelly import jelly, unjelly
>>> unjelly(jelly(Referenceable(), invoker=Broker()))
('remote', 1)
>>> 
```

This should be very easy to fix; in practice it doesn't cause many problems because banana doesn't differentiate between lists and tuples, so even though the tuple gets serialized and sent over a PB connection, it gets unserialized to a list and passes the check in the unjelly method.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="aruse's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|aruse set owner to aruse<br>aruse set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="aruse's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|aruse removed owner<br>aruse set status to `new`|
|-|-|

There are actually two bugs here:

  1. twisted.spread.flavors.Referenceable.jellyFor returns a tuple instead of a list during the jelly operation.

  2. The twisted.spread.pb.Broker class has no unserializingPerspective attribute, which raises an AtrributeError during the unjelly operation.

The patch above should fix both issues.

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy set owner to @djfroofy<br>@djfroofy set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_11"></a>|@djfroofy set owner to @exarkun<br>@djfroofy set status to `new`|
|-|-|

+1

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [31087]) Branching to 'implicit-jelly-555'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [31088]) Apply patch

refs [#10410](https://github.com/twisted/twisted/issues/10410)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [31089]) A few minor tweaks

  * Be more explicit in the docstring
  * name variables with camelCase instead of under_scores
  * Move imports to the top-level (or remove redundant ones)
  * lose the list instance check, it's overly specific

refs [#10410](https://github.com/twisted/twisted/issues/10410)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [31090]) news fragment
refs [#10410](https://github.com/twisted/twisted/issues/10410)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [31091]) Merge implicit-jelly-555

Author: aruse
Reviewer: djfroofy
Fixes: [#10410](https://github.com/twisted/twisted/issues/10410)

Change `Referenceable` to return a list from its `jellyFor` method so that
the result can be unjellied without travelling through banana.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_17"></a>|@markrwilliams commented|
|-|-|

In changeset ec42d7209d98f36b89b114520558d34a4034a44f
```
#!CommitTicketReference repository="" revision="ec42d7209d98f36b89b114520558d34a4034a44f"
Merge pull request #555 from twisted/8852-dist3-rm

Author: hawkowl
Reviewer: markrwilliams
Fixes: #8852

Move _dist3 into _setup
```
